### PR TITLE
New spider scaffold/pull request generator

### PR DESCRIPTION
**Pull request by CloCkWeRX**
Very tentative version of https://github.com/alltheplaces/alltheplaces/issues/9404

- Adds a command generator, so that you can generate two steps for "Make branch, spider template, open it in an editor" and "Run, commit, push to your fork"
- Adds templates for JSON Blob, Structured Data, WP Go Maps Spider variants. I chose to start here to avoid overlap with the autospidergen branch; but I can see the two approaches converging.
- Passing more variables to the spider for say, wikidata or other flags

What's missing?
- Chaining together these tools so that *if* you detect, say, structured data; it suggests you generate a spider using the template or autospidergen

### CODE REVIEW

File: templates/spiders/structured_data_spider_sitemap.tmpl [hunk 1]
```diff
@@ -0,0 +1,16 @@
+from scrapy.spiders import SitemapSpider
+
+from locations.structured_data_spider import StructuredDataSpider
+
+
+class $classname(SitemapSpider, StructuredDataSpider):
+    name = "$name"
+    #item_attributes = {"brand_wikidata": "Add if known"}
```
Comment (CloCkWeRX) [inline templates/spiders/structured_data_spider_sitemap.tmpl:8]:
```suggestion
    #item_attributes = {"brand": $name, "brand_wikidata": "Add if known"}
```

File: templates/spiders/json_blob.tmpl [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+class BurgermeisterSpider(WpGoMapsSpider):
+    name = "burgermeister"
+    item_attributes = {
+        "brand_wikidata": "Q116382535",
+        "brand": "Burgermeister",
+    }
+    allowed_domains = [
+        "burgermeister.com",
+    ]
```
Comment (iandees) [inline templates/spiders/json_blob.tmpl:33]: Is this supposed to be here?

Comment (CloCkWeRX) [reply templates/spiders/json_blob.tmpl]: Fixed

**Comment by CloCkWeRX** [comment]
@davidhicks thoughts on this alongside your branch? I feel like this is good for "I know there is a spider that is common" (jsonblob, structureddata, crawl, Sitemap, basic); though arguably we could go make a bunch of templates for each main Storefinder; for autodetect not working.

**Comment by davidhicks** [comment]
A few initial thoughts:
- Can template names be abbreviated e.g. "sds_crawl" instead of having to type "structured_data_spider_crawl"?
- Templates can/should easily include type hinting on all methods and class attributes?
- Includes "extras" in "item_attributes" with a dummy category that users would have to replace?

**Comment by panasonic-mafia** [comment]
I don't have access to a laptop now to check again, but when i tried to add custom templates last time, scrapy stopped to read standard templates...
https://github.com/scrapy/scrapy/issues/3680

**Comment by CloCkWeRX** [comment]
> Suggest to just transfer the default template file to new TEMPLATES_DIR directory.
Done

**Comment by CloCkWeRX** [comment]
Note to self:
https://github.com/scrapy/scrapy/pull/6470

**Comment by Cj-Malone** [comment]
> * Includes "extras" in "item_attributes" with a dummy category that users would have to replace?

I'm not sure we should encourage this. I know we do it sometimes, but I don't think it's ideal.

**Comment by davidhicks** [comment]
> > * Includes "extras" in "item_attributes" with a dummy category that users would have to replace?
> 
> I'm not sure we should encourage this. I know we do it sometimes, but I don't think it's ideal.

I thought this was encouraged (by mass additions of `{"extras": Categories.WHATEVER.value}` to `item_attributes` in the past approx.  6 months? It allows differentiation between two brands of the same name in the same country that have different primary OSM tags. For example, `Brand123` may be used for both `Categories.FUEL_STATION` and `Categories.SHOP_CONVENIENCE` within the same country. I note this is a somewhat rare situation though!

**Comment by Cj-Malone** [comment]
> > > * Includes "extras" in "item_attributes" with a dummy category that users would have to replace?
> > 
> > 
> > I'm not sure we should encourage this. I know we do it sometimes, but I don't think it's ideal.
> 
> I thought this was encouraged (by mass additions of `{"extras": Categories.WHATEVER.value}` to `item_attributes` in the past approx. 6 months? It allows differentiation between two brands of the same name in the same country that have different primary OSM tags. For example, `Brand123` may be used for both `Categories.FUEL_STATION` and `Categories.SHOP_CONVENIENCE` within the same country. I note this is a somewhat rare situation though.

It works for some cases, because I updated the pipeline to treat `extras` in `item_attributes` in a special way.

Eg

```python
from typing import Any

from scrapy import Spider
from scrapy.http import Response

from locations.categories import Categories, Extras, apply_yes_no
from locations.items import Feature


class Brand123Spider(Spider):
    name = "brand_123"
    item_attributes = {"extras": Categories.FUEL_STATION.value}
    start_urls = ["data:,"]

    def parse(self, response: Response, **kwargs: Any) -> Any:
        item = Feature(ref="123")
        apply_yes_no(Extras.SHOWERS, item, True)
        yield item
```
```python
{'extras': {'@source_uri': 'data:,',
            '@spider': 'brand_123',
            'amenity': 'fuel',
            'shower': 'yes'},
 'ref': 'abc'}
```

However in other cases, not the pipeline for example, it doesn't work as expected. eg:

```python
class BrandABCSpider(Spider):
    name = "brand_abc"
    start_urls = ["data:,"]

    def parse(self, response: Response, **kwargs: Any) -> Any:
        item = Feature(ref="abc")
        apply_yes_no(Extras.SHOWERS, item, True)
        item.update(Brand123Spider.item_attributes)
        yield item
```
```python
{'extras': {'@source_uri': 'data:,',
            '@spider': 'brand_abc',
            'amenity': 'fuel'},
 'ref': 'abc'}
```

There is also the possibility that we change how categories work at some point in the future, eg, a flat value instead of dicts (Like was recommended), this would make checking if a poi was a specific category much easier. Using `apply_category` in spider means we could update it without editing individual spiders, just the shared code.

**Comment by davidhicks** [comment]
Any thoughts on inserting an `ATPSpider` subclass that sits between Scrapy's `Spider` and all ATP spider classes that:
- Adds attributes: `brand`, `brand_wikidata`, `operator`, `operator_wikidata` where at least one brand or one operator is required.
- Adds attribute `top_level_tag` that accepts a `Categories.XYZ` dictionary. Required for all spiders.
- Adds attribute `locale` that accepts a `Locale.XYZ` class (will share soon: this is a branch that I've been working on to consolidate all translations for hours, "closed" labels, dictionary keys, etc into language-specific classes). Required for all spiders.

That way we'd largely remove the need for spiders to set `item_attributes` at all. The only time that `apply_category` would generally get used is the spiders which retrieve a variety of features such as convenience stores, fuel stations, etc all jumbled together.

File: locations/commands/genspider.py [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+            "module": module,
+            "name": name,
+            "url": url,
+            "domain": extract_domain(url),
+            "classname": f"{capitalized_module}Spider",
+        }
+        # Flatten the parameters and merge with the other variables
+        return tvars | self.parameters
+
+    # TODO: If https://github.com/scrapy/scrapy/pull/6470 is merged, much of this can be removed
```
Comment (CloCkWeRX) [inline locations/commands/genspider.py:120]: This did get merged, need to wait for release or point it git version and remove this method

**Pull request merged**